### PR TITLE
Main changes for kube/cluster-init.sh for multi-node cluster handling

### PR DESCRIPTION
**Pull request by naiming-zededa**
- create the cluster-utils.sh in pkg/kube for a number of functions
- get EdgeNodeClusterStatus from 'zedkube' for cluster status, cluster nodeip, cluster prefix, etc
- a background process to monitor the cluster mode changes
- implement cluster mode change logic and operations for change from single-node to cluster, or from cluster back to single-node
- save kube /var/lib at single-node first time setup for later use when the mode changes back from cluster mode to single node
- display eve-release string for get node OS-IMAGE field
- this patch tested still works in kubevirt single-node case.
- some of the external_boot_img changes in the patch is from Andrew's diff.

### CODE REVIEW

File: pkg/kube/cluster-init.sh [hunk 1]
```diff
@@ -1,141 +1,109 @@
 #!/bin/sh
 #
-# Copyright (c) 2023-2024 Zededa, Inc.
+# Copyright (c) 2024 Zededa, Inc.
```
Comment (eriknordmark) [inline pkg/kube/cluster-init.sh:3]: You shouldn't remove the old year.

File: pkg/kube/cluster-init.sh [hunk 2]
```diff
@@ -255,30 +250,61 @@ longhorn_post_install_config() {
...
+        if [ ! -d /var/lib/cni/bin ] || [ ! -d /opt/cni/bin ]; then
+                copy_cni_plugin_files
+        fi
+        # for now, always copy to get the latest
+
+        # start the k3s server now
+        nohup /usr/bin/k3s server --config "$k3s_config_file" &
+
+        # remember the k3s start time
+        k3s_last_start_time=$(date +%s)
```
Comment (eriknordmark) [inline pkg/kube/cluster-init.sh:285]: Did you intend to use this variable?

File: pkg/kube/cluster-init.sh (hunk 1 above)
Comment (naiming-zededa) [reply pkg/kube/cluster-init.sh]: fixed. not sure how this was modified.

File: pkg/kube/cluster-init.sh (hunk 2 above)
Comment (naiming-zededa) [reply pkg/kube/cluster-init.sh]: removed, no need for this in the latest change.

File: pkg/kube/cluster-init.sh [hunk 3]
```diff
@@ -295,62 +318,45 @@ check_start_containerd() {
...
                 # This import happens once per reboot
-                if ctr -a /run/containerd-user/containerd.sock image import /etc/external-boot-image.tar; then
-                        eve_external_boot_img_tag=$(cat /run/eve-release)
-                        eve_external_boot_img=docker.io/lfedge/eve-external-boot-image:"$eve_external_boot_img_tag"
-                        import_tag=$(tar -xOf /etc/external-boot-image.tar manifest.json | jq -r '.[0].RepoTags[0]')
-                        ctr -a /run/containerd-user/containerd.sock image tag "$import_tag" "$eve_external_boot_img"
-
-                        logmsg "Successfully installed external-boot-image $import_tag as $eve_external_boot_img"
-                        rm -f /etc/external-boot-image.tar
+                import_name_tag=$(tar -xOf /etc/external-boot-image.tar manifest.json | jq -r '.[0].RepoTags[0]')
```
Comment (andrewd-zededa) [inline pkg/kube/cluster-init.sh:345]: This is in my PR #4363

**Review by andrewd-zededa — CHANGES_REQUESTED** [review]
This PR is including some changes from my PR, also yetus cleanup needed.

File: pkg/kube/cluster-init.sh [hunk 4]
```diff
@@ -498,49 +797,91 @@ if [ ! -f /var/lib/all_components_initialized ]; then
         fi
 
         if [ ! -f /var/lib/longhorn_initialized ]; then
-                wait_for_item "longhorn"
-                logmsg "Installing longhorn version ${LONGHORN_VERSION}"
-                apply_longhorn_disk_config "$HOSTNAME"
-                lhCfgPath=/var/lib/lh-cfg-${LONGHORN_VERSION}.yaml
-                if [ ! -e $lhCfgPath ]; then
-                        curl -k https://raw.githubusercontent.com/longhorn/longhorn/${LONGHORN_VERSION}/deploy/longhorn.yaml > "$lhCfgPath"
+                if [ ! -f /var/lib/longhorn_installing ]; then
```
Comment (andrewd-zededa) [inline pkg/kube/cluster-init.sh:800]: This is also from #4363

File: pkg/kube/cluster-init.sh [hunk 5]
```diff
@@ -498,49 +797,91 @@ if [ ! -f /var/lib/all_components_initialized ]; then
...
+                        fi
+                        if ! grep -q 'create-default-disk-labeled-nodes: true' "$lhCfgPath"; then
+                                sed -i '/  default-setting.yaml: |-/a\    create-default-disk-labeled-nodes: true' "$lhCfgPath"
+                        fi
+                        kubectl apply -f "$lhCfgPath"
+                        touch /var/lib/longhorn_installing
                 fi
-                if ! grep -q 'create-default-disk-labeled-nodes: true' "$lhCfgPath"; then
-                        sed -i '/  default-setting.yaml: |-/a\    create-default-disk-labeled-nodes: true' "$lhCfgPath"
+                lhStatus=$(kubectl -n longhorn-system get daemonsets -o json | jq '.items[].status | .numberReady==.desiredNumberScheduled' | tr -d '\n')
```
Comment (andrewd-zededa) [inline pkg/kube/cluster-init.sh:814]: This is also from #4363

**Review by andrewd-zededa — COMMENTED** [review]
A few more notes

File: pkg/kube/cluster-init.sh (hunk 3 above)
Comment (naiming-zededa) [reply pkg/kube/cluster-init.sh:345]: Yeh, we need to coordinate to see which one merge in first, then modify accordingly.

Comment (andrewd-zededa) [reply pkg/kube/cluster-init.sh:345]: @naiming-zededa Can you remove my changes from your PR?

Comment (naiming-zededa) [reply pkg/kube/cluster-init.sh]: Removed with my latest patch.

**Comment by deitch** [comment]
Looks like it needs to be rebased.

**Comment by naiming-zededa** [comment]
> Looks like it needs to be rebased.

Hi @deitch rebased in the update.
hi @andrewd-zededa, i have resolved some conflicts, in particular this:
```
        if ! external_boot_image_import; then
                continue
        fi
```
please check to see if this is in the intended place.

**Comment by eriknordmark** [comment]
> hi @andrewd-zededa, i have resolved some conflicts, in particular this:

@andrewd-zededa can you respond to this and take another look?

File: pkg/kube/cluster-init.sh [hunk 6]
```diff
@@ -3,143 +3,110 @@
...
+                kubectl create namespace eve-kube-app
+        fi
+        logmsg "Apply multus-daemonset-new.yaml"
+        if ! kubectl apply -f /etc/multus-daemonset-new.yaml > /dev/null 2>&1; then
+                logmsg "Apply Multus, has failed, jump out now"
                 return 1
         fi
         logmsg "Done applying Multus"
-        ln -s /var/lib/cni/bin/multus /var/lib/rancher/k3s/data/current/bin/multus
+        link_multus_into_k3s
```
Comment (andrewd-zededa) [inline pkg/kube/cluster-init.sh:109]: `link_multus_into_k3s` is not defined yet, its in an upcoming PR for pkg/kube/cluster-update.sh

**Review by andrewd-zededa — CHANGES_REQUESTED** [review]
I think the original conflicts/overlap are resolved, just a small fix for a function not yet defined.

File: pkg/kube/cluster-init.sh (hunk 6 above)
Comment (andrewd-zededa) [reply pkg/kube/cluster-init.sh:109]: I can submit a quick PR to add a skeleton cluster-update.sh with just this defined to get this moving again.

Comment (andrewd-zededa) [reply pkg/kube/cluster-init.sh:109]: I'm curious why shellcheck didn't find this

Comment (andrewd-zededa) [reply pkg/kube/cluster-init.sh:109]: I created #4445 to define the missing function

File: pkg/kube/cluster-init.sh (hunk 4 above)
Comment (andrewd-zededa) [reply pkg/kube/cluster-init.sh]: this looks resolved

**Review by eriknordmark — APPROVED** [review]
LGTM

**Pull request merged**